fetch/cache.go

```diff
@@ -42,17 +51,15 @@ func NewHashPeersCache(size int) *HashPeersCache {
 
 // Add adds peer to a hash.
 func (hpc *HashPeersCache) Add(hash types.Hash32, peer p2p.Peer) {
-	hpc.mu.Lock()
-	defer hpc.mu.Unlock()
-
 	peers, exists := hpc.get(hash)
+	hpc.mu.Lock()
```

Comment:
this is a race condition. and the code is getting complicated for simply getting a list of peers. you now have 3 levels of get (GetList, Get, get)

how about just change Get() to return a list instead of a map? both callers of Get() can work with list.